lib/experimental/sync/class-gutenberg-http-signaling-server.php

```diff
@@ -0,0 +1,343 @@
+		}
+		return $topics_to_subscribers;
+	}
+
+	/**
+	 * Receives a $topics_to_subscribers data-structure and an array of topics,
+	 * and returns a new $topics_to_subscribers data-structure where the current subscriber is not subscribed to the topics.
+	 *
+	 * @access private
+	 * @internal
+	 *
+	 * @param array $topics_to_subscribers  Topics to subscribers data-structure.
+	 * @param array $topics                 An array of topics e.g: array( 'doc1', 'doc2' ).
+	 */
+	private static function unsubscribe_from_topics( $topics_to_subscribers, $topics ) {
+		foreach ( $topics as $topic ) {
+			if ( $topics_to_subscribers[ $topic ] ) {
+				$topics_to_subscribers[ $topic ] = array_diff( $topics_to_subscribers[ $topic ], array( static::$subscriber_id ) );
+			}
+		}
+		return $topics_to_subscribers;
+	}
+
+	/**
+	 * Handles a wp_ajax signaling server request of client that is performing an operation.
+	 * An operation can be a ping to say the client is alive, sending a message, or subscribing/unscribing to a set of topics.
+	 *
+	 * @param array $message                 An array of topics e.g: array( 'doc1', 'doc2' ).
+	 */
+	private static function handle_message_operation( $message ) {
+		$fd_topics_subscriber = fopen( static::$topics_to_subscribers_path, 'c+' );
+		if( ! $fd_topics_subscriber ) {
+			die( 'Could not open required file.' );
+		}
+		flock( $fd_topics_subscriber, LOCK_EX );
+		$topics_to_subscribers = static::get_contents_from_file_descriptor( $fd_topics_subscriber );
+
+		switch ( $message['type'] ) {
+			case 'subscribe':
+				static::save_contents_to_file_descriptor( $fd_topics_subscriber, static::subscribe_to_topics( $topics_to_subscribers, $message['topics'] ) );
+				break;
+			case 'unsubscribe':
+				static::save_contents_to_file_descriptor( $fd_topics_subscriber, static::unsubscribe_from_topics( $topics_to_subscribers, $message['topics'] ) );
+				break;
+			case 'publish':
+				$fd_subscriber_messages = fopen( static::$subscriber_to_messages_path, 'c+' );
+				if( ! $fd_subscriber_messages ) {
+					die( 'Could not open required file.' );
+				}
+				flock( $fd_subscriber_messages, LOCK_EX );
+				$subscriber_to_messages = static::get_contents_from_file_descriptor( $fd_subscriber_messages );
+				$topic                  = $message['topic'];
+				$receivers              = $topics_to_subscribers[ $topic ];
+				if ( $receivers ) {
+					$message['clients'] = count( $receivers );
+					foreach ( $receivers as $receiver ) {
+						if ( ! $subscriber_to_messages[ $receiver ] ) {
+							$subscriber_to_messages[ $receiver ] = array();
+						}
+						$subscriber_to_messages[ $receiver ][] = $message;
+					}
+					static::save_contents_to_file_descriptor( $fd_subscriber_messages, $subscriber_to_messages );
+				}
+				flock( $fd_subscriber_messages, LOCK_UN );
+				fclose( $fd_subscriber_messages );
+				break;
+			case 'ping':
```

Comment:
Not for this PR, but a thought for later. I wonder if we could invert this, so it's the server who sends pings to the clients to keep the connection alive via server-sent events. I presume this could be less resource intensive for the server.

Reply:
I don't think the server can actually send an event to a client because the client is the one who initiates the connection. However, I believe we can completely eliminate the need for pinging messages. Instead, the client can request messages, and if the server responds with a normal retry, the client can assume that the server is alive. Right now, my intention is to maintain a close connection with y-webrtc.